enforcer/pucontext/pucontext.go

```diff
@@ -1,37 +1,223 @@
+func (p *PUContext) ID() string {
+	return p.id
+}
+
+// ManagementID returns the management ID
+func (p *PUContext) ManagementID() string {
+	return p.managementID
+}
+
+// Type return the pu type
+func (p *PUContext) Type() constants.PUType {
+	return p.puType
+}
+
+// Identity returns the indentity
+func (p *PUContext) Identity() *policy.TagStore {
+	return p.identity
+}
+
+// IP returns the IP of the PU
+func (p *PUContext) IP() string {
+	return p.ip
+}
+
+// Mark returns the PU mark
+func (p *PUContext) Mark() string {
+	return p.mark
+}
+
+// Ports returns the PU ports
+func (p *PUContext) Ports() []string {
+	return p.ports
+}
+
+// Annotations returns the annotations
+func (p *PUContext) Annotations() *policy.TagStore {
+	return p.annotations
+}
+
+// SearchAcceptTxRules returns the accept Tx Rules
+func (p *PUContext) SearchAcceptTxRules(claims *policy.TagStore) (int, interface{}) {
+	return p.acceptTxRules.Search(claims)
+}
+
+// SearchRejectTxRules searches the reject rules for a policy match
+func (p *PUContext) SearchRejectTxRules(claims *policy.TagStore) (int, interface{}) {
+	return p.rejectTxRules.Search(claims)
+}
+
+// SearchAcceptRcvRules searches the accept rules for a policy match
+func (p *PUContext) SearchAcceptRcvRules(claims *policy.TagStore) (int, interface{}) {
+	return p.acceptRcvRules.Search(claims)
+}
+
+// SearchRejectRcvRules returns the reject receive rules
+func (p *PUContext) SearchRejectRcvRules(claims *policy.TagStore) (int, interface{}) {
+	return p.rejectRcvRules.Search(claims)
+}
+
+// RetrieveCachedExternalFlowPolicy returns the policy for an external IP
+func (p *PUContext) RetrieveCachedExternalFlowPolicy(id string) (interface{}, error) {
+	return p.externalIPCache.Get(id)
+}
+
+// NetworkACLPolicy retrieves the policy based on ACLs
+func (p *PUContext) NetworkACLPolicy(packet *packet.Packet) (*policy.FlowPolicy, error) {
+	return p.networkACLs.GetMatchingAction(packet.SourceAddress.To4(), packet.DestinationPort)
+}
+
+// ApplicationACLPolicy retrieves the policy based on ACLs
+func (p *PUContext) ApplicationACLPolicy(packet *packet.Packet) (*policy.FlowPolicy, error) {
+	return p.applicationACLs.GetMatchingAction(packet.SourceAddress.To4(), packet.SourcePort)
+}
+
+// CacheExternalFlowPolicy will cache an external flow
+func (p *PUContext) CacheExternalFlowPolicy(packet *packet.Packet, plc interface{}) {
+	p.externalIPCache.AddOrUpdate(packet.SourceAddress.String()+":"+strconv.Itoa(int(packet.SourcePort)), plc)
+}
+
+// GetProcessKeys returns the cache keys for a process
+func (p *PUContext) GetProcessKeys() (string, []string) {
+	return p.mark, p.ports
+}
+
+// SynServiceContext returns synServiceContext
+func (p *PUContext) SynServiceContext() []byte {
+	p.RLock()
+	defer p.RUnlock()
+	return p.synServiceContext
+}
+
+// UpdateSynServiceContext updates the synServiceContext
+func (p *PUContext) UpdateSynServiceContext(synServiceContext []byte) {
+
+	p.Lock()
+	defer p.Unlock()
```

Comment:
Defer is not cheap. Here (and in other functions) you can avoid it

Reply:
sure.